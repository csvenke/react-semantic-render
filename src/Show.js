import PropTypes from "prop-types";

import render from "./utils/render";

function Show(props) {
  return props.when ? render(props) : null;
}

Show.propTypes = {
  when: PropTypes.bool.isRequired,
  children: PropTypes.node,
  render: PropTypes.func,
};

export default Show;
