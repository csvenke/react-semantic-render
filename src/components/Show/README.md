### Example usage

Simple example with children.

```jsx
const showButton = true;

const App = () => (
  <Show when={showButton}>
    <button>Click me!</button>
  </Show>
);

<App />;
```

Simple example with render. <br />
_Use this when you don't want your content evaluated unless an condition is true._

```jsx
let obj = undefined;

obj = {
  label: 'Click me!',
};

const App = () => (
  <Show
    when={!!obj}
    render={() => (
      <button>
        <span>{obj.label}</span>
      </button>
    )}
  />
);

<App />;
```
