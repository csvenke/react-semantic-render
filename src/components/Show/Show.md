### How to use

```jsx static
import React from 'react';
import { Show } from 'react-semantics';
```

#### With render prop
##### Output

```jsx
const data = faker.fake("{{lorem.words}}");

<Show 
  when={true}
  render={() => (
    <div>{data}</div>
  )}
/>
```


#### With children
##### Output

```jsx
const data = faker.fake("{{lorem.words}}");

<Show when={true}>
  <div>{data}</div>
</Show>
```