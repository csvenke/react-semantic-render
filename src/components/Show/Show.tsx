import * as PropTypes from 'prop-types';
import * as React from 'react';

import { isEmptyChildren } from '../../utils';

export interface IShowProps {
  /** Conditional statement. Will be cast to true/false value  */
  when: any;

  /** Shorthand for primary content. */
  render?: () => React.ReactNode;

  /** Primary content. */
  children?: React.ReactNode;
}

/**
 * Renders primary content if `when` equals true.
 *
 * @example
 *
 * <Show when={2 === 1 + 1}>
 *  <div>render me</div>
 * </Show>
 *
 * <Show
 *  when={2 === 1 + 1}
 *  render={() => <div>render me</div>}
 * />
 */
const Show: React.SFC<IShowProps> = ({ when, render, children }) => {
  if (!!when) {
    if (children && !isEmptyChildren(children)) {
      return React.Children.only(children);
    }

    if (render) {
      return <React.Fragment>{render()}</React.Fragment>;
    }
  }

  return null;
};

Show.propTypes = {
  children: PropTypes.node,
  render: PropTypes.func,
  when: PropTypes.any,
};

export default Show;
