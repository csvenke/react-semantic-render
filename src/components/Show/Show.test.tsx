import { shallow } from 'enzyme';
import * as React from 'react';

import Show from '.';

const input = <div>render me</div>;
const output = '<div>render me</div>';

describe('with children', () => {
  test('should return primary content from children if when equals true', () => {
    const element = <Show when>{input}</Show>;
    const wrapper = shallow(element);
    expect(wrapper.html()).toEqual(output);
  });

  test('should return null if when equals false', () => {
    const element = <Show when={false}>{input}</Show>;
    const wrapper = shallow(element);
    expect(wrapper.html()).toEqual(null);
  });
});

describe('with render', () => {
  test('should return primary content from render if when equals true', () => {
    const element = <Show when render={() => input} />;
    const wrapper = shallow(element);
    expect(wrapper.html()).toEqual(output);
  });

  test('should return null if when equals false', () => {
    const element = <Show when={false} render={() => input} />;
    const wrapper = shallow(element);
    expect(wrapper.html()).toEqual(null);
  });
});

describe('with children and render', () => {
  const renderInput = <div>render</div>;
  const childrenInput = <div>children</div>;
  const childrenOutput = '<div>children</div>';

  test('should return primary content from children and ignore render if when equals true', () => {
    const element = (
      <Show when render={() => renderInput}>
        {childrenInput}
      </Show>
    );
    const wrapper = shallow(element);
    expect(wrapper.html()).toEqual(childrenOutput);
  });

  test('should return null if when equals false', () => {
    const element = (
      <Show when={false} render={() => renderInput}>
        {childrenInput}
      </Show>
    );
    const wrapper = shallow(element);
    expect(wrapper.html()).toEqual(null);
  });
});

describe('without children and render', () => {
  test('should return null', () => {
    const element = <Show when />;
    const wrapper = shallow(element);
    expect(wrapper.html()).toEqual(null);
  });
});
