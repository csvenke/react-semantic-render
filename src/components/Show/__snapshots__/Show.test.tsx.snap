// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with children and render should return null if when equals false 1`] = `null`;

exports[`with children and render should return primary content from children and ignore render if when equals true 1`] = `
<div>
  children
</div>
`;

exports[`with children should return null if when equals false 1`] = `null`;

exports[`with children should return primary content from children if when equals true 1`] = `
<div>
  render me
</div>
`;

exports[`with render should not evaluate render if when is undefined 1`] = `null`;

exports[`with render should return null if when equals false 1`] = `null`;

exports[`with render should return primary content from render if when equals true 1`] = `
<div>
  render me
</div>
`;

exports[`without children and render should return null 1`] = `null`;
